src/ucp/rndv/rndv_put.c

```diff
@@ -126,26 +126,40 @@ ucp_proto_rndv_put_common_atp_send(ucp_request_t *req, ucp_lane_index_t lane)
 {
     const ucp_proto_rndv_put_priv_t *rpriv = req->send.proto_config->priv;
     ucp_proto_rndv_put_atp_pack_ctx_t pack_ctx;
+    ucs_status_t status;
+
+    /* Make sure the sum of ack_size field in all ATP messages we send will not
+       exceed request length, since each ATP message has to acknowledge at least
+       one byte. */
+    ucs_assertv(req->send.rndv.put.atp_count <= req->send.state.dt_iter.length,
+                "atp_count=%u length=%zu", req->send.rndv.put.atp_count,
+                req->send.state.dt_iter.length);
+    if (req->send.rndv.put.atp_count == req->send.state.dt_iter.length) {
+        return UCS_OK;
+    }
 
     pack_ctx.req = req;
 
     /* When we need to send multiple ATP messages: each will acknowledge 1 byte,
        except the last ATP which will acknowledge the remaining payload size.
        This is simpler than keeping track of how much was sent on each lane */
-    if (lane == ucs_ilog2(rpriv->atp_map)) {
+    if (req->send.rndv.put.atp_count == (rpriv->atp_num_lanes - 1)) {
         pack_ctx.ack_size = req->send.state.dt_iter.length -
-                            rpriv->atp_num_lanes + 1;
-        if (pack_ctx.ack_size == 0) {
-            return UCS_OK; /* Skip sending 0-length ATP */
-        }
+                            req->send.rndv.put.atp_count;
     } else {
         pack_ctx.ack_size = 1;
     }
 
-    return ucp_proto_am_bcopy_single_send(req, UCP_AM_ID_RNDV_ATP, lane,
-                                          ucp_proto_rndv_put_common_pack_atp,
-                                          &pack_ctx, sizeof(ucp_rndv_ack_hdr_t),
-                                          0);
+    status = ucp_proto_am_bcopy_single_send(req, UCP_AM_ID_RNDV_ATP, lane,
+                                            ucp_proto_rndv_put_common_pack_atp,
+                                            &pack_ctx,
+                                            sizeof(ucp_rndv_ack_hdr_t), 0);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ++req->send.rndv.put.atp_count;
+    return UCS_OK;
```

Comment:
one line shorter:
```suggestion
    if (status == UCS_OK) {
        ++req->send.rndv.put.atp_count;      
    }

    return status;
```

Reply:
the suggestion is shorter but the current implementation is according to code style where error flow is checked first